### ISSUE TITLE
voice.makeRests() fills in missing notes in tuplets incorrectly

### ISSUE THREAD

**Issue by gregchapman-dev** [cuthbertLab/music21#1237]
**music21 version**

`7.2.1`   

**Problem summary**

If I put a single triplet eighth-note rest at the end of a 3-quarter-note voice (leaving a starting gap of 11 triplet eighth notes) and then call makeRests, it inserts a rest with duration.tuplets = (<music21.duration.Tuplet 12/11/16th>,) instead of what I would expect: duration.tuplets = (<music21.duration.Tuplet 3/2/eighth>,).   The actual duration.quarterLength is correct (11/3ql), but I don't understand that weird tuplet choice (12 actual 16th notes in the space of 11 normal 16th notes).

**Steps to reproduce**

```
from fractions import Fraction
import music21 as m21
restVoice = m21.stream.Voice()
secondNote = m21.note.Rest()
secondNote.duration = m21.duration.Duration(Fraction(1, 3))
restVoice.insert(Fraction(11, 3), secondNote)
restVoice.makeRests(inPlace=True, hideRests=True)
firstNote = restVoice.notesAndRests[0]
print('firstNote = ', firstNote)
print('firstNote tuplets = ', firstNote.duration.tuplets)
print('secondNote = ', secondNote)
print('secondNote tuplets = ', secondNote.duration.tuplets)
```

**Expected vs. actual behavior**

**More information**

**Comment by mscuthbert** [comment]
What happens if you give a time signature?

**Comment by gregchapman-dev** [comment]
Same thing. (The real case has one)

**Comment by gregchapman-dev** [comment]
I take it back!

If I put a time signature in the Voice itself, this works correctly.

In the real case, the time signature occurred earlier in the score.  But the Voice is not yet inserted when I do the makeRests(), so it doesn’t know.  I will try putting it in the score first, to see if that helps.

Greg

**Comment by gregchapman-dev** [comment]
Ah.  So it only works if the time signature is in the Voice itself.  Having the voice inserted into the measure (which has an implied time signature from a previous measure) doesn’t help.

So I do need a fix here.

Greg

**Comment by jacobtylerwalls** [comment]
Most likely another manifestation of #904.

When you say this works with a time signature (albeit not in a real world scenario), what is the result? A half rest, quarter rest, and two triplet eighths? If so, I would have thought #992 was needed for that (speaking of which, the procedure described there is probably your best workaround for now).

**Comment by gregchapman-dev** [comment]
I was wrong, in several ways.  Nothing I have done so far actually makes a rest with a reasonable tuplet. I am investigating further.

Greg

**Comment by jacobtylerwalls** [comment]
Another very similar discussion at #572.

**Comment by gregchapman-dev** [comment]
I have tried a LOT of things now… Here’s the latest attempt, with everything but the kitchen sink thrown in.

The rest that makeRests creates is still that weird tuplet (12 notes in the space of 11).

Jacob, thanks for the pointer to #904.  I’ll try that procedure described there as a workaround now.
restVoice = m21.stream.Voice()
timesig = m21.meter.TimeSignature('4/4')
restVoice.insert(0, timesig)
secondNote = m21.note.Rest()
secondNote.duration = m21.duration.Duration(Fraction(1, 3))
restVoice.insert(Fraction(11, 3), secondNote)
restVoice.makeRests(inPlace=True,
                    hideRests=True,
                    fillGaps=True,
                    timeRangeFromBarDuration=True,
                    refStreamOrTimeRange=timesig)
firstNote = restVoice.notesAndRests[0]
print('firstNote = ', firstNote)
print('firstNote tuplets = ', firstNote.duration.tuplets)
print('secondNote = ', secondNote)
print('secondNote tuplets = ', secondNote.duration.tuplets)
print('')
<stdout>
firstNote =  <music21.note.Rest 11/3ql>
firstNote tuplets =  (<music21.duration.Tuplet 12/11/16th>,)
secondNote =  <music21.note.Rest 1/3ql>
secondNote tuplets =  (<music21.duration.Tuplet 3/2/eighth>,)

**Comment by gregchapman-dev** [comment]
OK, I tried using duration.TupletFixer, and it finds the right tupletGroup, but then fails to correct it, saying:

Crazy! 4.0 2.75 16/11

(once I uncommented that print, and fixed a typo in it)

I don’t blame it, I can’t get my head around what that first tuplet means, either.

Greg

**Comment by gregchapman-dev** [comment]
Here's my code for that use of TupletFixer:

```
from fractions import Fraction
import music21 as m21
print(m21.__version__)

restVoice = m21.stream.Voice()
timesig = m21.meter.TimeSignature('4/4')
#restVoice.insert(0, timesig)
secondNote = m21.note.Rest()
secondNote.duration = m21.duration.Duration(Fraction(1, 3))
restVoice.insert(Fraction(11, 3), secondNote)
restVoice.makeRests(inPlace=True,
                    hideRests=True) #,
                    #fillGaps=True,
                    #timeRangeFromBarDuration=True,
                    #refStreamOrTimeRange=timesig)
tf = m21.duration.TupletFixer(restVoice)
tupletGroups = tf.findTupletGroups()
for tg in tupletGroups:
    tf.fixBrokenTupletDuration(tg)

firstNote = restVoice.notesAndRests[0]
print('firstNote = ', firstNote)
print('firstNote tuplets = ', firstNote.duration.tuplets)
print('secondNote = ', secondNote)
print('secondNote tuplets = ', secondNote.duration.tuplets)
```

**Comment by jacobtylerwalls** [comment]
Thanks for looking into some workarounds.

I'll be closing as a duplicate of #572, but I went ahead and re-opened a PR to fix it (#1240), so I'm hoping you might test it out. It's not truly mergeable until we have a solution for #904, but I'll open something for that soon and figured this way testing/feedback could get going, if you have some. Thanks.
